src/main/java/htsjdk/tribble/gff/Gff3Feature.java

```diff
@@ -0,0 +1,417 @@
+
+    /**
+     * Gets set of parent features
+     * @return set of parent features
+     */
+    public Set<Gff3Feature> getParents() {return parents;}
+
+    /**
+     * Gets set of features for which this feature is a parent
+     * @return set of child features
+     */
+    public Set<Gff3Feature> getChildren() {return children;}
+
+    /**
+     * Get set of all features this feature descends from, through chains of Parent attributes.  Derives_From can be used to specify a particular inheritance path for this feature when multiple paths are available
+     * @return set of ancestor features
+     */
+    public Set<Gff3Feature> getAncestors() {
+        final List<Gff3Feature> ancestors = new ArrayList<>(parents);
+        for (final Gff3Feature parent : parents) {
+            ancestors.addAll(baseData.attributes.containsKey(DERIVES_FROM_ATTRIBUTE_KEY)? parent.getAncestors(baseData.attributes.get(DERIVES_FROM_ATTRIBUTE_KEY)) : parent.getAncestors());
+        }
+        return ancestors.stream().collect(Collectors.toSet());
+    }
+
+    private Set<Gff3Feature> getAncestors(final String derivingFrom) {
+        final List<Gff3Feature> ancestors = new ArrayList<>();
+        for (final Gff3Feature parent : parents) {
+            if (parent.getID().equals(derivingFrom) || parent.getAncestors().stream().anyMatch(f -> f.getID().equals(derivingFrom))) {
+                ancestors.add(parent);
+                ancestors.addAll(parent.getAncestors());
+            }
+        }
+        return ancestors.stream().collect(Collectors.toSet());
+    }
+
+    /**
+     * Get set of all features descended from this features, through chains of Parent attributes.  Derives_From can be used to specify a particular inheritance path for this feature when multiple paths are available
+     * @return set of descendents
+     */
+    public Set<Gff3Feature> getDescendents() {
+        final List<Gff3Feature> descendants = new ArrayList<>(children);
+        final Set<String> idsInLineage = new HashSet<>(Collections.singleton(baseData.ID));
+        idsInLineage.addAll(children.stream().map(Gff3Feature::getID).collect(Collectors.toSet()));
+        for(final Gff3Feature child : children) {
+            descendants.addAll(child.getDescendents(idsInLineage));
+        }
+        return descendants.stream().collect(Collectors.toSet());
+    }
+
+    private Set<Gff3Feature> getDescendents(final Set<String> idsInLineage) {
+        final List<Gff3Feature> decendants = new ArrayList<>();
+        final List<Gff3Feature> childrenToAdd = children.stream().filter(c -> c.getAttribute(DERIVES_FROM_ATTRIBUTE_KEY) == null ||
+                idsInLineage.contains(c.getAttribute(DERIVES_FROM_ATTRIBUTE_KEY))).
+                collect(Collectors.toList());
+        decendants.addAll(childrenToAdd);
+        idsInLineage.addAll(childrenToAdd.stream().map(Gff3Feature::getID).collect(Collectors.toSet()));
```

Comment:
The modification does need to happen, since the Derives_from attribute could be pointing to a feature at any higher generation.  So as we move down through the generation, we need to keep adding to the list of ids in the particular lineage.  However, I can make this clearer (and reduce the side effect) by making a new list updated to include the new ids and call with that, instead of modifying the input list.